src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+                    ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+                if ((len % 2) != 0)
+                    ERROR_OUT(BUFFER_ERROR, dtls13_cleanup);
+                suites.hashSigAlgoSz = len;
+                XMEMCPY(suites.hashSigAlgo, tlsx.elements + OPAQUE16_LEN,
+                        len);
+            }
+
+            /* Supported groups */
+            ret = TlsxFindByType(&tlsx, TLSX_SUPPORTED_GROUPS,
+                                 ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                ret = TLSX_SupportedCurve_Parse(ssl, tlsx.elements,
+                                                tlsx.size, 1, &parsedExts);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+            }
+
+            /* Key share */
+            ret = TlsxFindByType(&tlsx, TLSX_KEY_SHARE,
+                                 ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                ret = TLSX_KeyShare_Parse_ClientHello(ssl, tlsx.elements,
+                                          tlsx.size, &parsedExts);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+                haveKS = 1;
+#endif
+            }
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            /* Pre-shared key */
+            ret = TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (tlsx.size != 0) {
+                /* Let's just assume that the binders are correct here. We will
+                 * actually verify this in the stateful part of the processing
+                 * and if they don't match we will error out there anyway. */
+                byte modes;
+
+                ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                     ch->extension);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if (tlsx.size == 0)
+                    ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                                  client_hello, &modes);
+                if (ret != 0)
+                    goto dtls13_cleanup;
+                if ((modes & (1 << PSK_DHE_KE)) && !ssl->options.noPskDheKe) {
+                    if (!haveKS)
+                        ERROR_OUT(MISSING_HANDSHAKE_DATA, dtls13_cleanup);
+                    doKE = 1;
+                }
+                else {
+                    if ((modes & (1 << PSK_KE)) == 0)
+                        ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+                }
+            }
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+            if (usePSK) {
```

Comment:
Maybe I'm missing it, but when do we set `usePSK`?

Reply:
You're right. I forgot to set it. It should be after the `TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension)`.